src/trace/context.ts

```diff
@@ -238,6 +245,162 @@ export function readTraceFromSQSEvent(event: SQSEvent): TraceContext | undefined
   return;
 }
 
+export function readTraceFromSNSSQSEvent(event: SQSEvent): TraceContext | undefined {
+  if (event.Records && event.Records[0] && event.Records[0].body) {
+    try {
+      const parsedBody = JSON.parse(event.Records[0].body) as SNSMessage;
+      if (
+        parsedBody.MessageAttributes &&
+        parsedBody.MessageAttributes._datadog &&
+        parsedBody.MessageAttributes._datadog.Value
+      ) {
+        const traceData = JSON.parse(parsedBody.MessageAttributes._datadog.Value);
+        const traceID = traceData[traceIDHeader];
+        if (typeof traceID !== "string") {
+          return;
+        }
+        const parentID = traceData[parentIDHeader];
+        if (typeof parentID !== "string") {
+          return;
+        }
+        const sampledHeader = traceData[samplingPriorityHeader];
+        if (typeof sampledHeader !== "string") {
+          return;
+        }
+        const sampleMode = parseInt(sampledHeader, 10);
+
+        const trace = {
+          parentID,
+          sampleMode,
+          source: Source.Event,
+          traceID,
+        };
+        logDebug(`extracted trace context from SNS SQS event`, { trace, event });
+        return trace;
+      }
+    } catch (err) {
+      if (err instanceof Error) {
+        logError("Error parsing SNS SQS message trace data", err as Error);
+      }
+      return;
+    }
+  }
+}
+
+export function readTraceFromKinesisEvent(event: KinesisStreamEvent): TraceContext | undefined {
+  if (event.Records && event.Records[0] && event.Records[0].kinesis && event.Records[0].kinesis.data) {
+    try {
+      const parsedBody = JSON.parse(Buffer.from(event.Records[0].kinesis.data, "base64").toString("ascii")) as any;
+      if (parsedBody && parsedBody._datadog) {
+        const traceData = parsedBody._datadog;
+        const traceID = traceData[traceIDHeader];
+        if (typeof traceID !== "string") {
+          return;
+        }
+        const parentID = traceData[parentIDHeader];
+        if (typeof parentID !== "string") {
+          return;
+        }
+        const sampledHeader = traceData[samplingPriorityHeader];
+        if (typeof sampledHeader !== "string") {
+          return;
+        }
+        const sampleMode = parseInt(sampledHeader, 10);
+
+        const trace = {
+          parentID,
+          sampleMode,
+          source: Source.Event,
+          traceID,
+        };
+        logDebug(`extracted trace context from Kinesis event`, { trace, event });
+        return trace;
+      }
+    } catch (err) {
+      if (err instanceof Error) {
+        logError("Error parsing Kinesis message trace data", err as Error);
+      }
+      return;
+    }
+  }
+}
+
+export function readTraceFromEventbridgeEvent(event: EventBridgeEvent<any, any>): TraceContext | undefined {
+  if (event.detail && event.detail._datadog) {
+    try {
+      const traceData = event.detail._datadog;
```

Comment:
I don't think we anticipate having a `_datadog` key in the `event.detail`. The way it was implemented in `dd-trace-py`, we just have `x-datadog-trace-id`, etc.

Reply:
Ah I see, I'd been following the convention for sns/sqs where we stick a `_datadog` string map in the message attributes, but I can change it to this. But yes, we need stay consistent here.